tools-md/src/main/java/org/aya/literate/Literate.java

```diff
@@ -0,0 +1,126 @@
+// Copyright (c) 2020-2023 Tesla (Yinsen) Zhang.
+// Use of this source code is governed by the MIT license that can be found in the LICENSE.md file.
+package org.aya.literate;
+
+import kala.collection.immutable.ImmutableSeq;
+import org.aya.pretty.doc.Doc;
+import org.aya.pretty.doc.Docile;
+import org.aya.pretty.doc.Link;
+import org.aya.pretty.doc.Style;
+import org.aya.util.error.SourcePos;
+import org.jetbrains.annotations.Contract;
+import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
+
+/**
+ * @author hoshino
+ * @see LiterateConsumer
+ */
+public interface Literate extends Docile {
+  record Raw(@NotNull Doc toDoc) implements Literate {
+  }
+
+  record List(@NotNull ImmutableSeq<Literate> items, boolean ordered) implements Literate {
+    @Override public @NotNull Doc toDoc() {
+      return Doc.list(ordered, items.map(Literate::toDoc));
+    }
+  }
+
+  record HyperLink(@NotNull String href, @Nullable String hover,
+                   @NotNull ImmutableSeq<Literate> children) implements Literate {
+    @Override public @NotNull Doc toDoc() {
+      var child = Doc.cat(this.children().map(Literate::toDoc));
+      return Doc.hyperLink(child, Link.page(href), hover);
+    }
+  }
+
+  record Image(@NotNull String src, @NotNull ImmutableSeq<Literate> children) implements Literate {
+    @Override public @NotNull Doc toDoc() {
+      var child = Doc.cat(this.children().map(Literate::toDoc));
+      return Doc.image(child, Link.page(src));
+    }
+  }
+
+  record Math(boolean inline, @NotNull ImmutableSeq<Literate> children) implements Literate {
+    @Override public @NotNull Doc toDoc() {
+      var child = Doc.cat(this.children().map(Literate::toDoc));
+      return inline ? Doc.math(child) : Doc.mathBlock(child);
+    }
+  }
+
+  record Many(@Nullable Style style, @NotNull ImmutableSeq<Literate> children) implements Literate {
+    @Override public @NotNull Doc toDoc() {
+      var child = Doc.cat(this.children().map(Literate::toDoc));
+      return style == null ? child : Doc.styled(style, child);
+    }
+  }
+
+  class CommonCode implements Literate {
```

Comment:
Why so many code blocks?

Reply:
We just need one code block class. The derived markdown parser should tell the base parser "the interesting languages".
And when the base parser encounters a code block, check if the language is in that list, if so, parse it to the code block instance, otherwise, parse it to `Literate.Raw(Doc.plain(...)``

Reply:
The base parser probably doesn't know how to construct the inline code block for that language, or you are talking about the fenced code block?

Reply:
Yes. It's fenced code block.

Reply:
Why don't we parse all the code blocks to `Literate.CodeBlock`?

Reply:
I think we probably can

Reply:
This helps markdown-in-markdown if it is not interesting, see https://github.com/aya-prover/aya-dev/pull/952

Reply:
Kiva said she's tackle this herself